### PR TITLE
policy: Track source policy rules in MapStateEntry

### PR DESCRIPTION
**Pull request by gandro**
This commit allows us to track the policies for which a certain policy
map entry has been created.

It is implemented by copying over the `DerivedFromRules` from the
merged ingress/egress filters to the user-space representation of
the policy map state. Default entries such as `AllowAnyIngress`,
`AllowAnyEgress` and `AllowLocalHostIngress` are annotated with
artificial labels (of label source `reserved`).

The entries are copied over into `realizedPolicy` of each endpoint when
the policy maps are synced. This enables us to expose this information
in the newly introduced `GetRealizedPolicyRuleLabelsForKey` getter.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10293)

### CODE REVIEW

**Comment by gandro** [comment]
test-me-please

**Comment by aanm** [comment]
@gandro it seems to be a real failure

```
# github.com/cilium/cilium/pkg/policy
vet: pkg/policy/distillery_test.go:249:20: undeclared name: NoRedirectEntry
make[1]: *** [govet] Error 2
make[1]: Leaving directory `/home/travis/gopath/src/github.com/cilium/cilium'
make: *** [unit-tests] Error 2
```

File: pkg/labels/arraylist.go [hunk 1]
```diff
@@ -40,3 +40,16 @@ func (ls LabelArrayList) GetModel() [][]string {
 	}
 	return res
 }
+
+// Same returns true if the label arrays lists have the same label arrays in the same order.
+func (ls LabelArrayList) Same(b LabelArrayList) bool {
```
Comment (aanm) [inline pkg/labels/arraylist.go:45]: Could this be named `Equal`? Can you add unit tests for this? what should be the result of this function for a non-nil received with a nil argument?

File: pkg/endpoint/policy.go [hunk 2]
```diff
@@ -751,3 +751,21 @@ func (e *Endpoint) UpdateVisibilityPolicy(annoCB AnnotationsResolverCB) {
 	}
 	<-ch
 }
+
+// GetRealizedPolicyRuleLabelsForKey returns the list of policy rule labels
+// which match a given flow key (in host byte-order)
+func (e *Endpoint) GetRealizedPolicyRuleLabelsForKey(key policy.Key) (
```
Comment (aanm) [inline pkg/endpoint/policy.go:757]: This function was added but it seems unused. The reason I'm asking this is because this can be optimized by having a LRU map that checks if an entry exists in the that lru map and keeps returning an already deepcopied value of rules. Otherwise we are increasing the memory consumption by always returning a deepcopy of these rules.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/labels/arraylist.go (hunk 1 above)
Comment (gandro) [reply pkg/labels/arraylist.go:45]: I named it `Same` to be consistent with the function is also called `Same` on the `LabelArray`. It is called `Equal` on `Label`.

Thanks for pointing out the missing unit tests. I actually wrote unit-tests for this, but apparently forgot to add them to the commit.

File: pkg/endpoint/policy.go (hunk 2 above)
Comment (gandro) [reply pkg/endpoint/policy.go:757]: Yes, it's going to be used by code outside the Cilium repository. Maybe I can add a comment about this. 

But the LRU cache is a good input, I'll see if I can avoid having to `DeepCopy` everything.

File: pkg/labels/arraylist.go (hunk 1 above)
Comment (joestringer) [reply pkg/labels/arraylist.go:45]: (Should they both be called `Equals` / `Equals()`? `if foo.Same(bar) { ... }` doesn't read terribly easily..)

File: pkg/policy/mapstate.go [hunk 3]
```diff
@@ -68,22 +77,47 @@ type MapStateEntry struct {
...
 	ProxyPort uint16
+
+	// DerivedFromRules tracks the policy rules this entry derives from
+	DerivedFromRules labels.LabelArrayList
+}
+
+// NewPlaceholderEntry creates a map state entry. If redirect is true, the
+// caller is expected to replace the ProxyPort field before it is added to
+// the actual BPF map.
+func NewPlaceholderEntry(derivedFrom labels.LabelArrayList, redirect bool) MapStateEntry {
```
Comment (joestringer) [inline pkg/policy/mapstate.go:88]: nit: I found the function naming here abstract and confusing. If there is no redirect, this is not a placeholder but already the final map state entry that leads directly to allowing traffic on the datapath side. 

I would just call this something like `NewMapStateEntry()`, since it creates a structure of that type (whether fully populated or not).

File: pkg/policy/mapstate.go [hunk 4]
```diff
@@ -100,7 +134,12 @@ func (keys MapState) RedirectPreferredInsert(key Key, entry MapStateEntry) {
 // endpoint.
 func (keys MapState) DetermineAllowLocalhostIngress(l4Policy *L4Policy) {
 	if option.Config.AlwaysAllowLocalhost() {
-		keys[localHostKey] = MapStateEntry{}
+		keys[localHostKey] = MapStateEntry{
```
Comment (joestringer) [inline pkg/policy/mapstate.go:137]: nit: This should use the function above too, so that if it gets refactored further in future the compiler will point out any further required changes, rather than silently ignoring / zeroing the added fields.

Comment (joestringer) [reply pkg/policy/mapstate.go:137]: (Same for the other direct initializations)

**Review by joestringer — CHANGES_REQUESTED** [review]
Just a couple of minor nits.

**Comment by gandro** [comment]
Thanks for the reviews! Testing revealed that this version is incomplete, will re-open another PR once this is ready.
